### PR TITLE
Calidation generic type support

### PR DESCRIPTION
**Pull request by lisatassone**
PR for adding generics to the calidation library types. I have added this as a breaking change because it's currently enforcing you pass through a type (even though that can be any of course). If this is a problem, I can update it to default to any, but this is meant to be a more type safe version. Although the original is helpful to know what the signatures are for the libraries methods and interfaces, working with the data you pass through to the form is always untyped which for my purposes, is not ideal.

First PR, so please let me know if there is anything further I need to do. I have tested these in a production app.

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [X] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Review by ArrayKnight — CHANGES_REQUESTED** [review]
I think this is really great and exactly what is needed.

However, I think we should do this as a non-breaking change. Since Calidation v2 may or may not happen, it would be great if these types were compatible with v1.16

It's excellent that you're bringing in more of the Typescript standards, like replacing Dictionary with Record.

Also, are there features you're using that are specific to TS v3.8? I know one of my projects is currently stuck on v3.5... Would it be possible to make this round of updates remain compatible with v3.5?

**Comment by lisatassone** [comment]
> I think this is really great and exactly what is needed.
> 
> However, I think we should do this as a non-breaking change. Since Calidation v2 may or may not happen, it would be great if these types were compatible with v1.16
> 
> It's excellent that you're bringing in more of the Typescript standards, like replacing Dictionary with Record.
> 
> Also, are there features you're using that are specific to TS v3.8? I know one of my projects is currently stuck on v3.5... Would it be possible to make this round of updates remain compatible with v3.5?

Thanks for the review @ArrayKnight :) I don't think I'm using anything that can't mean the types can stay on 3.5. Version 3.8 is just what I tested it with. As for keeping it compatible, I can default the generic to `any` so users don't have to supply it and it basically has the same types as before. I will try to update this today. I will update the version of the types to v1.16.2 as well.

File: types/calidation/index.d.ts [hunk 1]
```diff
@@ -1,134 +1,159 @@
-// Type definitions for calidation 1.16
+// Type definitions for calidation 1.17
```
Comment (ArrayKnight) [inline types/calidation/index.d.ts:1]: Is it possible for this to remain 1.16?

File: types/calidation/index.d.ts [hunk 2]
```diff
@@ -1,134 +1,159 @@
...
-}
+export type Dirty<T extends object = any> = Record<keyof T, boolean>;
+
+export type Errors<T extends object = any> = Record<keyof T, string | null>;
 
-export type Dirty = Dictionary<boolean>;
+export type Fields<T extends object = any> = T;
 
-export type Errors = Dictionary<string | null>;
+export type Transform<T extends object = any> = Record<keyof T, (value: any) => any>;
```
Comment (ArrayKnight) [inline types/calidation/index.d.ts:16]: I think the transforms should accept any as the value but return the type of the field.

**Review by ArrayKnight — CHANGES_REQUESTED** [review]

File: types/calidation/index.d.ts (hunk 1 above)
Comment (lisatassone) [reply types/calidation/index.d.ts]: Seems as though it can

File: types/calidation/index.d.ts (hunk 2 above)
Comment (lisatassone) [reply types/calidation/index.d.ts]: Done. I've removed the additional transforms func def since it probably isn't capturing the intent of that feature.

**Review by ArrayKnight — APPROVED** [review]

**Comment by lisatassone** [comment]
Ready to merge

**Pull request merged**